src/v/cluster/scheduling/leader_balancer_greedy.h

```diff
@@ -90,42 +97,51 @@ class greedy_balanced_shards final : private leader_balancer_strategy {
     find_movement(const absl::flat_hash_set<raft::group_id>& skip) const final {
         const auto curr_error = error();
 
-        /*
-         * from: high load core
-         * to: less loaded core
-         */
+        // Consider each group from high load core, and record the reassignment
+        // involving the lowest load "to" core.
         for (const auto& from : boost::adaptors::reverse(_load)) {
             if (_muted_nodes.contains(from->first.node_id)) {
                 continue;
             }
-            for (const auto& to : _load) {
-                // skip cores on muted nodes
-                if (_muted_nodes.contains(to->first.node_id)) {
+
+            constexpr size_t load_unset = std::numeric_limits<size_t>::max();
+            size_t lowest_load = load_unset;
+            reassignment lowest_reassign{};
+
+            // Consider each group from high load core, and record the
+            // reassignment involving the lowest load "to" core.
+            for (const auto& group : from->second) {
+                if (skip.contains(group.first)) {
                     continue;
                 }
-                // consider group from high load core
-                for (const auto& group : from->second) {
-                    if (skip.contains(group.first)) {
+
+                // iterate over all the replicas and look for the lowest load
+                // shard in the replica list
+                for (const auto& to_shard : group.second) {
+                    auto load = _load_map.at(to_shard);
+                    if (likely(load >= lowest_load)) {
+                        // there is no point in evaluating this move, it is
+                        // worse than the best one we've found so far.
                         continue;
                     }
-                    /*
-                     * a valid move requires that the group's replica set
-                     * contain the less loaded `to` core.
-                     */
-                    const auto it = std::find(
-                      group.second.cbegin(), group.second.cend(), to->first);
-                    if (it == group.second.cend()) {
```

Comment:
just to make sure I'm understanding this: instead of checking if the move is valid explicitly (that the move stays within the replica set), you are only considering valid moves when looking for a minimum cost move?

Reply:
That's correct we iterate over the replica set, so we never consider a "to" node that is not in the replica set. Note that this means that we actually consider the current leader (i.e., a no-op move) which should never be selected because it won't improve the error: I believe the old algorithm did the same.

Reply:
Cool. I think this PR can merge unless you'd like to get more feedback.

Reply:
I think it's good to go, I'm clicking the button.

Reply:
@dotnwat - I am sorry, I rebased my branch and it dismissed your review. There are no changes to the commit contents, it was just a rebase. Can you approve again?

Reply:
sorry i abandoned this mention @travisdowns! i think this was about the time i went awol working on v22.1 release stuff :/